### ISSUE TITLE
Internationalization improvements

### ISSUE THREAD

**Issue by Chris1234567899** [flutter/flutter#18835]
In my humble opinion, the internationalization could need some improval. I am following the suggestions from the tutorial using arb files and even though it is working, I think, a lot could be simplified there and steps reduced.
I should mention this comes from the user perspective and I might not be aware of all the reasons why it is implemented this way. If you think it is alright as it is, please close this issue, nonetheless I want to mention some things first:

1. Writing for every translation a method in localizations.dart has some drawbacks:

-  It get's very messy for large apps with lots of translations
-  The naming convention is limited (something like "my-custom-translation" or "home.title" is not possible due to the hyphen or point)
-  Why not recommending a generic method (`String getTranslation(key){ return...)`)?
-  The translations cannot be grouped. What I mean is somthing like
```
{
   home-page:{
      title:"Home",
      heading:"A nice homepage",
      texts:{
         desc1:"Bla bla",
         desc2:"Bla2 bla 2"
      }
   },
   settings-page:{...}
}
```
is not (easily) possible at the moment. This however would increase a lot readability and help maintaining an overview on where which translation is used

2. The intl_translation commands could be reduced/simplified

3. Too many documents: At the moment we have the "same" information in 3 files: The localization.dart, the intl_*.arb and the messages_*.dart.  This seems error prone and unneccesarily hard to maintain.

The ngx-translate library for angular for example seems way more straight forward and I think a lot of developers would approve something similar for flutter.
Just my 5 cents, keep up the great work!

**Comment by vladaman** [comment]
I absolutely agree. Internationalization in dart/flutter is just pain. I wish I could use one of the following systems:

1. use Android system with resources and easy references
2. use .po files somehow, but that requires yet another editor
3. use external json files (probably the easiest to manage)

At the end, we manage about 15 translations and it has become very painful and unmanageable in dart.

**Comment by alan-knight** [comment]
1. What do you find messy with a large number of translations? We do have people using it with applications with very very large numbers of translations, but that's in the context of particular tools.

Is Flutter requiring all the translations being in a particular file? Or is that just a convention for starting with. With intl_translation there is no requirement that messages be in a particular file, just that you run the extractor over all the files that have messages you want translated.

A single function wouldn't work for messages that take parameters, plurals, genders, etc, or messages that require extra parameters (meaning, description). However, I think you could add one for the simple case.

      getTranslation(String key) => Intl.message('', key);

The general grouping mechanism would be to put related translations in a single class, or a single library. So it could be

```
         desc1: Mystuff.bla1,
         desc2: Mystuff.bla2
```

Generally you can't just have the strings directly embedded because translators require additional information. And I'm not sure what you're showing here. So you have e.g. a JSON format that defines your window, and you want to embed the strings directly there? That doesn't look like what I've seen of Flutter code.

2. Could you be more specific about the simplifications?

3. Two of those documents are automatically generated, so it shouldn't be too hard to manage. Generally the expected workflow is that you write the original sources in your Dart code, generate a translation file from that which can be sent to translators, and then get the results back and those can be used to generate Dart code with the appropriate translations in it. The included format is ARB, because that fully supports the ICU message semantics, and there's at least one translation service (Google Translator's Toolkit) that knows how to consume it and produce translations easily. But depending what translation tools you're using, things might vary quite a bit. If you're manually writing translations in ARB files, I can see how that would be quite awkward, and I would not recommend it.

**Comment by Chris1234567899** [comment]
Thanks for your answer. I try to be more specific:
1. For a medium or large app you easily have 1000+ translations, meaning several 1000 lines of code in the localization.dart file. Maintaining this is hard in terms of searching, and editing these methods, since it is not really structured, just a long bunch of methods, sometimes with similar names. I don't say it's impossible, just that it makes things unnecessary complex. Splitting it again in different files seems rather  to worsen it.
I agree on the parameterization/plural thing, this is something to think about. Maybe parameters could be added in a generic function as optional arguments?

About the classes I'm not sure what you mean.

My example was not from flutter, but from ngx-translate, which is using a simple JSON file for each language. It was only to show what I mean by grouping translations. You call a translation by it's key directly in the page and that's it - super straight forward.

2. is connected with 3. 
Assuming I am writing some widgets, I would like to add translations on the fly and not rerun the intl commands for every single translation to see my changes. If you read the flow you just described, you see that it's around three corners -> .dart -> .arb ->.dart...
The ARB format is perfectly fine, but why not reading the values directly from there through a generic function (with parameters if necessary)? In the end it's a json structure.

**Comment by freezy** [comment]
@alan-knight what you're describing sounds reasonable given there are decent tools that can handle `.arb` files. Under "decent" I understand that:

- the tool can *update* the source messages from an updated `.arb` file
- the tool can import *translated* messages from an already generated `.arb` file

These are the fundamental features to make a translation tool usable, because a) an application lives and gets updated with new messages, and b) we don't want to re-translate everything in case a translation already exists in the `.arb` format.

As far as I can tell, Google Translator's Toolkit does neither of those. It works somewhat well if you only need to translate one file one time, but as updates come in you can either start from scratch or edit the `.arb` files manually.

So what's going on? Did I miss a tool you guys have been using for maintaining translations of your Flutter apps? Am I too moronic to use the Translator Toolkit properly?

Could you maybe describe, given we have a source `.arb` file and many translated `.arb` files that are updated constantly, how a workflow with technically unsavvy translaters would work? What tools are the developers with "very very large numbers of translations" using?

**Comment by alan-knight** [comment]
re: Google Translator's Toolkit, see https://github.com/dart-lang/intl_translation/issues/44

**Comment by lclc** [comment]
Any progress on this?

All the problems have been outlined above. A multi-language Flutter app is not very convenient to maintain at the moment.

**Comment by HansMuller** [comment]
@lclc, @freezy, @lifenautjoe, @Chris1234567899, @vladaman We're considering providing a tool that generates a class that encapsulates all of an app's translations, from the app's ARB files: https://github.com/flutter/flutter/issues/41437

The overall goal is to simplify building and maintaining an international app. We'd appreciate your feedback on the proposal.

**Comment by jaumard** [comment]
@alan-knight  Google Translator's Toolkit is soon dead and gone https://support.google.com/translatortoolkit/answer/9462068 

I think that being able to export to other format than arb is quite important if you want to work with online platform like PoEditor (and anything like it)

**Comment by jamesblasco** [comment]
Hello, these past few days I have been playing with  intl_translation  and the future version of gen_l10n.dart to build an **online arb editor** that creates the app_strings.dart and messages_*.dart files.

It is very early stage but I would love to hear some feedback before starting making stable and usable. Hope you like it
https://jamesblasco.github.io/arb_editor/master/#/

**Comment by shihaohong** [comment]
@jamesblasco i poked around at the app and it looks like a great tool! One quick thing i noticed was that it's great for if you had a template arb file to start, but may not take in to account existing translations you may already have for some messages. 

For example, if your template was 'en_US' and it had 25 messages, and you were done with 13/25 messages for 'es' and 'am', you would have to re-enter all 13 messages if there isn't a way to save it on the editor. Maybe provide an option to pass in other arb files as well so that it can populate already translated messages? 

Also, I assume the more stable version might take into account the additional formatting for numbers, currencies, and DateTime.

**Comment by jamesblasco** [comment]
Yeah i think being able to import arb files after the project created and being able to delete languages and messages are a must.
Btw now you can already drag multiple arb files at the same time and it will create a project with the all translations and not just the template

The numbers/concurrencies/DateTime support is already added to the roadmap 🙂

Thank you so much for the feedback!! 
This is a three-day work so I hope the first stable version to be more helpful.

**Comment by aleksakrstic** [comment]
You can use Flutter Intl plugin for IDE to get rid of manual work whenever you want to add/update/delete string keys in ARB files:

- VS Code: https://marketplace.visualstudio.com/items?itemName=localizely.flutter-intl
- Android Studio: https://plugins.jetbrains.com/plugin/13666-flutter-intl

**Comment by sketchbuch** [comment]
@Chris1234567899 

I agree it is messy and complicated and as a react developer I find it very verbose compared to say react react-i18next

> * It get's very messy for large apps with lots of translations
> * The translations cannot be grouped. What I mean is somthing like

Instead of your grouping ideas I have starte making new classes of translations for each page for large apps with many trandlations, this helped me a lot. But it is still messy with all the xxx.of(context).xxx everywhere. I assign the xxx.of(context) to a variable in the build method so and then it makes the whole widget tree more readable especially if there are a lot of translations used.

Hopefully though things will be improved!

**Comment by shihaohong** [comment]
Since the gen_l10n tool is now ready (See flutter.dev/go/i18n-user-guide), I will now close this issue. The tool allows developers to start from `arb` files and generates the l10n Dart files that Flutter uses for localization and message lookup. 

If there are any problems in the tool or if there was some problem in this thread that wasn't properly addressed by the tool, please file a separate issue, thank you!

**Comment by jbojcic1** [comment]
@shihaohong do you know when we can expect this to be available in a stable release? I am not able to figure it out from the milestones page.
